sql/src/main/java/io/crate/execution/dsl/phases/HashJoinPhase.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * Licensed to Crate under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.  Crate licenses this file
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.execution.dsl.phases;
+
+import io.crate.exceptions.UnsupportedFeatureException;
+import io.crate.execution.dsl.projection.Projection;
+import io.crate.expression.symbol.Symbol;
+import io.crate.expression.symbol.Symbols;
+import io.crate.planner.node.dql.join.JoinType;
+import io.crate.types.DataType;
+import io.crate.types.DataTypes;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.UUID;
+
+public class HashJoinPhase extends JoinPhase {
+
+    private final List<Symbol> leftJoinConditionInputs;
+    private final List<Symbol> rightJoinConditionInputs;
+
+    private final Collection<DataType> leftOutputTypes;
+    private final int leftSize;
```

Comment:
maybe this could be renamed to `leftNumRowsEstimate` or something like that, to make clear it's not accurate